src/cargo/sources/git/mod.rs

```diff
@@ -6,5 +6,66 @@ mod source;
 mod utils;
 
 pub mod fetch {
+    use crate::core::features::GitoxideFeatures;
+    use crate::Config;
+
+    /// The kind remote repository to fetch.
+    #[derive(Debug, Copy, Clone)]
+    pub enum RemoteKind {
+        /// A repository belongs to a git dependency.
+        GitDependency,
+        /// A repository belongs to a git dependency, and due to usage of checking out specific revisions we can't
+        /// use shallow clones.
+        GitDependencyForbidShallow,
+        /// A repository belongs to a Cargo registry.
+        Registry,
+    }
+
+    #[derive(Debug, Clone)]
+    pub enum History {
+        Shallow(gix::remote::fetch::Shallow),
+        Unshallow,
+    }
+
+    impl From<History> for gix::remote::fetch::Shallow {
+        fn from(value: History) -> Self {
+            match value {
+                History::Unshallow => gix::remote::fetch::Shallow::undo(),
+                History::Shallow(how) => how,
+            }
+        }
+    }
+
+    impl RemoteKind {
+        /// Obtain the kind of history we would want for a fetch from our remote knowing if the target repo is already shallow
+        /// via `repo_is_shallow` along with gitoxide-specific feature configuration via `config`.
+        pub(crate) fn to_history(&self, repo_is_shallow: bool, config: &Config) -> History {
+            let has_feature = |cb: &dyn Fn(GitoxideFeatures) -> bool| {
+                config
+                    .cli_unstable()
+                    .gitoxide
+                    .map_or(false, |features| cb(features))
+            };
+            let how = if repo_is_shallow {
+                if matches!(self, RemoteKind::GitDependencyForbidShallow) {
+                    return History::Unshallow;
+                } else {
+                    gix::remote::fetch::Shallow::NoChange
```

Comment:
I think the gist of this advice is that unshallowing shallow clones intentionally or accidentally incurs high cost on the server as it bypasses caches. 

Unintentionally unshallowing is something that depends on branch configuration *and* refspecs (assuming the client code supports shallow), and the fact that `cargo` limits refspecs for fetches means that unintentional unshallowing would be the exception, not the rule. That could happen nonetheless when the history is rewritten past the shallow boundary for example.

For the crates index, I think this is really something to think about as users would run into this very situation every couple of months. But when they do, the server would traverse only a small history, so the cost incurred is not on the worst end of the spectrum.

Now that we have an implementation to show for, we might even pass it maybe along with an analysis to GitHub to be sure, and I think it's worth seeing this on a timescale as well. We are talking about *maybe* a year from now when this code might become available by default once `git2` is phased out. And by then, the HTTP index is probably already cemented as default so the amount of clones we't actually see like this would be greatly reduced.

Addressed in 5716fe521cd49fe71d17723fd6ea0292ce626533.

Reply:
Just for the record. Here are some concern of shallow clones from GitHub Infra Team in the past.

* https://github.com/CocoaPods/CocoaPods/issues/4989#issuecomment-193772935
* https://github.com/Homebrew/brew/pull/9383

We will continue the conversation on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/.60gitoxide.60.20integration.20updates) and post a summary here when we make progess.